Unify method in which file content is read across all resources

eramoto opened the pull request:
There are the similar issues as PR #2302. Almost resources return false positives when a file does not exist or is not read.

eramoto committed:
Create file-check functionality into utility file

There are the similar issues as PR #2302. Almost resources return false
positives when a file does not exist or is not read.

Signed-off-by: ERAMOTO Masaya <eramoto.masaya@jp.fujitsu.com>

eramoto committed:
Replace to file-check functionality

Signed-off-by: ERAMOTO Masaya <eramoto.masaya@jp.fujitsu.com>

adamleff commented on lib/utils/read_file.rb:
```diff
@@ -0,0 +1,24 @@
+# encoding: utf-8
+# author: ERAMOTO Masaya
+
+module ReadFile
```
Our other utilities are named as nouns... ObjectTraverser, CommentParser... while yours is named like a verb/action. Can I suggest changing this to `FileReader` instead (and alter the file name as well to `file_reader.rb`)?

adamleff reviewed (changes requested):
@eramoto This is a supremely great improvement. Excellent use of a mixin to remove repetitive code and make it easier for future resource authors to read file content.

I'd like to request a name change of the module to make it consistent with our other mixins. Otherwise, I think this is fantastic.

adamleff commented on lib/utils/file_reader.rb:
```diff
@@ -0,0 +1,24 @@
+    if raw_content.nil?
+      raise Inspec::Exceptions::ResourceSkipped, "Can't read file: #{path}"
+    end
+    if raw_content.empty?
```
This feels like something that should be left up to each resource. What if a file is expected to be empty?

eramoto replied:
I thought that users expected that any description exists in a file if they specified its path. But they are permitted to read it even if it is empty so they may expected that. Thus I will fix above.

adamleff replied:
One suggestion: leave this here but make it optional. The caller can specify an option, like so:

`read_file_contents(‘/tmp/empty’, allow_empty: true)`

adamleff replied:
Let me know if you like that idea and if you need help implementing it! 🙂

eramoto replied:
I apologize that I was unaware of your helpful advice.
I implemented to split and wrap it but I will rework to implement the caller having the optional argument.

eramoto replied:
I changed `read_file_content` as described later.
The optional argument is set at false as default since I think `allow_empty: true` should be explicitly specified in case allowing to read a empty file.

```Ruby
--- a/lib/utils/file_reader.rb
+++ b/lib/utils/file_reader.rb
@@ -2,7 +2,7 @@
 # author: ERAMOTO Masaya

 module FileReader
-  def read_file_content(path)
+  def read_file_content(path, allow_empty = false)
     # these are currently ResourceSkipped to maintain consistency with the resource
     # pre-refactor (which used skip_resource). These should likely be changed to
     # ResourceFailed during a major version bump.
@@ -15,7 +15,8 @@ module FileReader
     if raw_content.nil?
       raise Inspec::Exceptions::ResourceSkipped, "Can't read file: #{path}"
     end
-    if raw_content.empty?
+
+    if !allow_empty && raw_content.empty?
       raise Inspec::Exceptions::ResourceSkipped, "File is empty: #{path}"
     end

```

By the way, this PR does not allow dh_params and x509_certificate resources to read a empty file since they always raise a exception if a file is empty.

eramoto committed:
Remove workaround of sshd_config

Removes the workaround of sshd_config since Travis CI fails due to a bug
of dev-sec/ssh-baseline and the PR #100 will fix it.

Signed-off-by: ERAMOTO Masaya <eramoto.masaya@jp.fujitsu.com>

eramoto committed:
Fix dh_params and x509_certificate resources

If a file is empty, OpenSSL::PKey::DH and OpenSSL::X509::Certificate have
raised an exception and have skipped the inspection. Thus x509_certificate
and dh_params resources are not allowed to read a empty file.

Signed-off-by: ERAMOTO Masaya <eramoto.masaya@jp.fujitsu.com>

eramoto committed:
to_s of shadow expects filters is not nil

Signed-off-by: ERAMOTO Masaya <eramoto.masaya@jp.fujitsu.com>

jquick commented on lib/resources/auditd_conf.rb:
```diff
@@ -32,17 +36,7 @@ def read_params
-        skip_resource "Can't read file '#{@conf_path}'"
-        return @params = {}
-      end
+      content = read_file_content(@conf_path)
```
May just want to use `@content` here so we don't have to call down again. Another option is to just remove `@content` from the init block and let it be read here.

eramoto replied:
Hi @jquick , thanks for pointing them out.

When they do not call the read_file_content method in the init block, i.e. removing it from their init blocks, they do not skip the test and do report it as failure if a target file is not found since their to_s methods seem to interrupt the exception from methods.

Thus I will remove the extra call and use `@content` of the init block in the method.

jquick commented on lib/resources/bond.rb:
```diff
@@ -13,20 +14,22 @@ class Bond < FileResource
     def read_content
       # parse the file
-      @content = @file.content
+      @content = read_file_content(@path, allow_empty: true)
```
We should be able to ditch this line as it will be set in init.

jquick commented on lib/resources/inetd_conf.rb:
```diff
@@ -32,19 +36,8 @@ def method_missing(name)
 
     def read_params
       return @params if defined?(@params)
+      content = read_file_content(@conf_path)
```
Same here, lets just use `@content` in this method as we already have the info. Or remove the `@content` from the init.

jquick commented on lib/resources/limits_conf.rb:
```diff
@@ -24,19 +28,7 @@ def method_missing(name)
-        skip_resource "Can't read file \"#{@conf_path}\""
-        return @params = {}
-      end
+      content = read_file_content(@conf_path)
```
Same here.

jquick commented on lib/resources/login_def.rb:
```diff
@@ -36,19 +40,7 @@ def method_missing(name)
-        skip_resource "Can't read file \"#{@conf_path}\""
-        return @params = {}
-      end
+      content = read_file_content(@conf_path)
```
Same here.

jquick commented on lib/resources/ntp_conf.rb:
```diff
@@ -34,17 +38,7 @@ def to_s
-        skip_resource "Can't read file \"#{@conf_path}\""
-        return @params = {}
-      end
+      content = read_file_content(@conf_path)
```
Same here.

jquick reviewed (changes requested):
Hey @eramoto ! Fantastic work here. Just a few notes.

eramoto committed:
Use init block variable in methods

Signed-off-by: ERAMOTO Masaya <eramoto.masaya@jp.fujitsu.com>

jquick reviewed (approved):
Thanks @eramoto !
